RFC-0001-prsestissimo-metrics.md

```diff
@@ -0,0 +1,251 @@
+
+2. **SUM:** Tracks the sum of the inserted values. This STAT type can be assigned to metrics which track aggregate of Operator specific counters. For instance, Velox defines a metric called `velox.spill_input_bytes`. This is a global metric tracking the total spill input bytes across all operators at that instant. Each `Operator` that needs to spill to disk maintains a `Spiller` instance which also has a `SpillStats` member. This instance of SpillStats only tracks Operator specific events. Every time an Operator’s spill stat is updated, in this case, the spill_input_bytes, a Operator thread level counter, is updated as well. At the time of reporting, the sum of `spill_input_bytes` across all threads is gathered and only report the change since the last time aggregated. Note that thread level spill_input_bytes increase with time, they are not reset when an Operator finishes. So, the delta computed at global level across all threads is positive.
+
+3. **AVG:** Tracks the average of the inserted values. This stat type can be assigned to counters that grow or decay with time. For instance System or user CPU utilization, System or User memory usage.
+4. **RATE**: Tracks the sum of the inserted values per second. As of now, there are no references to this `StatType` in the Velox repo.
+4. **HISTOGRAMS:** This stat type is a summary of an event over a period of time. Histograms consists of buckets as keys which represent ranges for a metric and values are the counts representing the number of times the metric was recorded in that range. For instance, `kCounterHttpClientPrestoExchangeOnBodyBytes` is a histogram with min value 0 and max value 1000000. The size of each bucket is is 1000, this Histogram will have `(max -min)/bucket-size = (1000000 - 0)/1000 = 1000` buckets. The parameters following the max value are {50,90,95,99,100} which
+indicate that this Histogram metric tracks 50th, 90th etc. percentiles respectively.
+```
+EFINE_HISTOGRAM_METRIC(
+      kCounterHttpClientPrestoExchangeOnBodyBytes,
+      1000,
+      0,
+      1000000,
+      50,
+      90,
+      95,
+      99,
+      100);
+```
+### Current Reporting Flow
+
+#### Event driven metrics reporting (Blocking Call)
+
+Most of the Velox metrics are reported when a specific event occurs. It’s not periodic like Prestissimo counters. For instance, when memory is reclaimed from a query pool, we try to maintain a counter that is incremented each time when this event occurs:
+```
+uint64_t MemoryReclaimer::run(
+    const std::function<uint64_t()>& func,
+    Stats& stats) {
+  stats.reclaimedBytes += reclaimedBytes;
+  RECORD_HISTOGRAM_METRIC_VALUE(
+      kMetricMemoryReclaimExecTimeMs, execTimeUs / 1'000);
+  RECORD_METRIC_VALUE(kMetricMemoryReclaimedBytes, reclaimedBytes);
+  return reclaimedBytes;
+}
+```
+This type of reporting could occur in the execution path of a query and has potential to adversely affect query latency.
```

Comment:
Do we have perf data to show that these counters can impact query latency? We are updating an integer variable and I am not sure it would impact query latency as much.